### PR TITLE
feat!: update Autosuggest component to better support freeform and selected values

### PR DESCRIPTION
**Pull request by brian-smith-tcril**
## Description

This updates the Autosuggest component to provide the consumer with:
* The value entered in the textbox by the user
* The selected value (as it is displayed in the dropdown)
* The id of the selected option

This also allows the consumer to determine if the Autosuggest component should be in an error state when
* No text has been entered
* Text has been entered, but not selection has been made
* An external validation check has failed - "customError"

The consumer will provide error messages for the appropriate error states

BREAKING CHANGE: value of Autosuggest component is now an object instead of a string
BREAKING CHANGE: Autosuggest component now uses onChange instead of onSelected
BREAKING CHANGE: Autosuggest component now takes in different error messages for value/selection required, and custom errors

### Deploy Preview

https://deploy-preview-2899--paragon-openedx.netlify.app/components/form/form-autosuggest/

## Merge Checklist

* [ ] If your update includes visual changes, have they been reviewed by a designer? Send them a link to the Netlify deploy preview, if applicable.
* [ ] Does your change adhere to the documented [style conventions](https://github.com/openedx/paragon/blob/master/docs/decisions/0012-css-styling-conventions)?
* [ ] Do any prop types have missing descriptions in the Props API tables in the documentation site (check deploy preview)?
* [ ] Were your changes tested using all available themes (see theme switcher in the header of the deploy preview, under the "Settings" icon)?
* [ ] Were your changes tested in the `example` app?
* [ ] Is there adequate test coverage for your changes?
* [ ] Consider whether this change needs to reviewed/QA'ed for accessibility (a11y). If so, please add `wittjeff` and `adamstankiewicz` as reviewers on this PR.

## Post-merge Checklist

* [ ] Verify your changes were released to [NPM](https://www.npmjs.com/package/@edx/paragon) at the expected version.
* [ ] If you'd like, [share](https://github.com/openedx/paragon/discussions/new?category=show-and-tell) your contribution in [#show-and-tell](https://github.com/openedx/paragon/discussions/categories/show-and-tell).
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

### CODE REVIEW

File: src/Form/FormAutosuggest.jsx
Comment (monteri) [inline src/Form/FormAutosuggest.jsx:391]: I assume that there should be `valueRequiredErrorMessageText` provided if `valueRequired` equals `true`. Please, consider:

```
import { requiredWhen } from '../utils/propTypes';

...

valueRequiredErrorMessageText: requiredWhen(PropTypes.string, 'valueRequired'),
```

And the same for `selectionRequired` and `selectionRequiredErrorMessageText`, `customError` and `customErrorMessageText`.

Another option, if no message is provided and there should only be a red box to indicate the error, I would remove the `x` under the input, otherwise it will appear incomplete.image

File: src/Form/form-autosuggest.mdx
Comment (PKulkoRaccoonGang) [inline src/Form/form-autosuggest.mdx:83]: It seems to me that this `mdx` file should have 2 indent spaces (currently 4). This can be seen by looking at other `mdx` files of the [Form component](https://github.dev/openedx/paragon/blob/716a6ddd932d61a5b5ab594e8e525c3dfffe20ab/src/Form/form-autosuggest.mdx).

File: src/Form/tests/FormAutosuggest.test.jsx
Comment (PKulkoRaccoonGang) [inline src/Form/tests/FormAutosuggest.test.jsx:71]: [question]: Does it make sense to add additional comments to the details in the test file? It seems to me that we already have these descriptions on the page of the component itself.

**Review by PKulkoRaccoonGang — COMMENTED** [review]
@brian-smith-tcril leave some suggestions, please take a look

File: src/Form/FormAutosuggest.jsx
Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx]: Good catch! I removed it.

File: src/Form/FormAutosuggest.jsx
Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx]: I added some spaces and comments in here, hopefully that clears things up a bit.

File: src/Form/form-autosuggest.mdx
Comment (brian-smith-tcril) [reply src/Form/form-autosuggest.mdx]: I fixed the indentation (by copying the jsx from the mdx file into a temp.jsx file and running lint fix)

File: src/Form/tests/FormAutosuggest.test.jsx
Comment (brian-smith-tcril) [reply src/Form/tests/FormAutosuggest.test.jsx]: I was following the pattern of the existing ones, but I agree they felt redundant. I've removed them.

File: src/Form/FormAutosuggest.jsx
Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx]: @monteri I have updated these to use `requiredWhen`

https://github.com/openedx/paragon/blob/f8a4e5ee118199936c3ff6653e3dfb389d12fffa/src/Form/FormAutosuggest.jsx#L421-L430

File: src/Form/form-autosuggest.mdx
Comment (adamstankiewicz) [inline src/Form/form-autosuggest.mdx:90]: nit: consider using `Stack` or CSS utility class to give some vertical whitespace between the `Form.Control` and the `Button`.

image

File: src/Form/form-autosuggest.mdx
Comment (adamstankiewicz) [inline src/Form/form-autosuggest.mdx:25]: nit: perhaps prefix these boolean state variables with `isValueRequired`, `isSelectionRequired`, `hasCustomValidation`.

File: src/Form/FormAutosuggest.jsx
Comment (adamstankiewicz) [inline src/Form/FormAutosuggest.jsx:178]: [nit/suggestion] May be a good idea to include a comment here or elsewhere documenting the use of `useImperativeHandle` to only include `updateErrorStateAndErrorMessage` on its exposed methods. That way, future contributions will have some guidance on how to keep this maintained such that methods intended to be exposed are intentionally included here? 🤔

File: src/Form/FormAutosuggest.jsx
Comment (adamstankiewicz) [inline src/Form/FormAutosuggest.jsx:45]: nit: `isDowndownExpanded`

File: src/Form/FormAutosuggest.jsx
Comment (adamstankiewicz) [inline src/Form/FormAutosuggest.jsx:55]: nit: could this be simplified to:

```jsx
const [displayValue, setDisplayValue] = useState(value?.userProvidedText || '');
```

File: src/Form/FormAutosuggest.jsx
Comment (adamstankiewicz) [inline src/Form/FormAutosuggest.jsx:97]: nit: to avoid `// eslint-disable-next-line no-shadow`, consider renaming `children` once deconstructed (i.e., `const { children: childChildren, ... } = child.props;` to avoid the naming conflict with the parent `children`.

File: src/Form/FormAutosuggest.jsx
Comment (adamstankiewicz) [inline src/Form/FormAutosuggest.jsx:18]: [nit/suggestion/consideration] I'm wondering if it might make sense to abstract some of this logic into custom React hook(s) to keep the component code itself more focused on the UI presentation. For example, a custom hook(s) could return the callback functions to the component for use, having been defined in the hook(s).

**Review by adamstankiewicz — APPROVED** [review]

File: src/Form/form-autosuggest.mdx
Comment (brian-smith-tcril) [reply src/Form/form-autosuggest.mdx:92]: addressed here https://github.com/openedx/paragon/compare/f8a4e5ee118199936c3ff6653e3dfb389d12fffa..9310aa04d988eba1ee524cf920e5ad54dd39590f#diff-3ce694685ee1cc2b274e86b07ff39240bff96ddf3862d4c5366d0f24ad863d38R81-R91

File: src/Form/form-autosuggest.mdx
Comment (brian-smith-tcril) [reply src/Form/form-autosuggest.mdx]: addressed throughout https://github.com/openedx/paragon/compare/f8a4e5ee118199936c3ff6653e3dfb389d12fffa..9310aa04d988eba1ee524cf920e5ad54dd39590f

File: src/Form/FormAutosuggest.jsx
Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx]: added a comment here https://github.com/openedx/paragon/compare/f8a4e5ee118199936c3ff6653e3dfb389d12fffa..9310aa04d988eba1ee524cf920e5ad54dd39590f#diff-96e13c1a773ec8d218ea5667d3dec6c8f6eeadcb00c81584f50d6c3d4d52135fR170-R171

If you have another suggestion for a comment I'm more than happy to change it!

File: src/Form/FormAutosuggest.jsx
Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx]: addressed https://github.com/openedx/paragon/compare/f8a4e5ee118199936c3ff6653e3dfb389d12fffa..9310aa04d988eba1ee524cf920e5ad54dd39590f#diff-96e13c1a773ec8d218ea5667d3dec6c8f6eeadcb00c81584f50d6c3d4d52135fR45

File: src/Form/FormAutosuggest.jsx
Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx]: addressed https://github.com/openedx/paragon/compare/f8a4e5ee118199936c3ff6653e3dfb389d12fffa..9310aa04d988eba1ee524cf920e5ad54dd39590f#diff-96e13c1a773ec8d218ea5667d3dec6c8f6eeadcb00c81584f50d6c3d4d52135fR49

File: src/Form/FormAutosuggest.jsx
Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx]: addressed https://github.com/openedx/paragon/compare/f8a4e5ee118199936c3ff6653e3dfb389d12fffa..9310aa04d988eba1ee524cf920e5ad54dd39590f#diff-96e13c1a773ec8d218ea5667d3dec6c8f6eeadcb00c81584f50d6c3d4d52135fR90

File: src/Form/FormAutosuggest.jsx
Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx:18]: I'd like to think about this one a bit. I worry that *too* much abstraction might lead to consumers using the component in ways that feel broken, but that could likely be resolved by providing good examples.

Maybe it makes sense to break this out into a separate issue to do some discovery on it?

Comment (adamstankiewicz) [reply src/Form/FormAutosuggest.jsx:18]: The suggestion to abstract some of the logic shouldn't really affect consumers? The custom hooks wouldn't be exposed to consumers; just for use internally within the component.

Generally when a components implementation reaches several hundred plus lines that feels like the component might have a bit too much going on where it could be simplified through the use of hooks to improve readability, etc. 

Totally fine to defer and consider. Just a suggestion to make the component a bit more maintainable, where it makes sense.

Comment (brian-smith-tcril) [reply src/Form/FormAutosuggest.jsx:18]: > The custom hooks wouldn't be exposed to consumers; just for use internally within the component.

Gotcha, I misinterpreted the comment.

> Totally fine to defer and consider. Just a suggestion to make the component a bit more maintainable, where it makes sense.

That sounds good. It's definitely a good idea, but I think it makes sense to hold off until after getting this PR merged.

**Review by adamstankiewicz — APPROVED** [review]

**Pull request merged**
